File: packages/fpl-api-client/python/fpl_client.py
"""
fpl-api-client · packages/fpl-api-client/python/fpl_client.py
==============================================================
Unified HTTP client for the official Fantasy Premier League API.

SOURCE:  Extracted and generalised from:
  - fpl-video-repurposer/build_fpl_kb.py  (fetch_json, BOOTSTRAP_URL, FIXTURES_URL,
                                           build_master_squad, build_next_fixture_map)

REPLACES (do NOT delete originals until migration is approved):
  - fpl-video-repurposer/build_fpl_kb.py  → import from this module instead

CONSUMERS AFTER MIGRATION:
  - fpl-video-repurposer/build_fpl_kb.py
  - captaincy-ml pipelines that need live FPL data
  - fpl-platform/pipelines/build_player_registry.py
"""

from __future__ import annotations

import time
from typing import Any

import requests

# ---------------------------------------------------------------------------
# Endpoint constants
# ---------------------------------------------------------------------------

BOOTSTRAP_URL = "https://fantasy.premierleague.com/api/bootstrap-static/"
FIXTURES_URL = "https://fantasy.premierleague.com/api/fixtures/?event={gameweek}"
ELEMENT_SUMMARY_URL = "https://fantasy.premierleague.com/api/element-summary/{player_id}/"

# Default HTTP settings
_DEFAULT_TIMEOUT = 30
_RETRY_ATTEMPTS = 3
_RETRY_BACKOFF = 2.0  # seconds, doubles each attempt


# ---------------------------------------------------------------------------
# Low-level fetch helper
# ---------------------------------------------------------------------------

def fetch_json(url: str, timeout: int = _DEFAULT_TIMEOUT) -> Any:
    """Fetch a URL and return parsed JSON.

    Raises:
        requests.HTTPError: on non-2xx responses after retries
        requests.ConnectionError: on network failure after retries

    SOURCE: fpl-video-repurposer/build_fpl_kb.py::fetch_json (lines 22-26)
    """
    last_exc: Exception | None = None
    for attempt in range(1, _RETRY_ATTEMPTS + 1):
        try:
            resp = requests.get(url, timeout=timeout)
            resp.raise_for_status()
            return resp.json()
        except (requests.HTTPError, requests.ConnectionError) as exc:
            last_exc = exc
            if attempt < _RETRY_ATTEMPTS:
                time.sleep(_RETRY_BACKOFF * attempt)
    raise last_exc  # type: ignore[misc]


# ---------------------------------------------------------------------------
# Bootstrap data (players + teams + gameweek events)
# ---------------------------------------------------------------------------

def get_bootstrap() -> dict[str, Any]:
    """Return the full FPL bootstrap-static response.

    Contains:
      - elements      → all players with stats, status, ownership, xG/xA
      - teams         → team names, codes, strengths, fixtures difficulty
      - events        → gameweek events (deadline, is_finished, etc.)
      - element_types → position definitions (1=GKP, 2=DEF, 3=MID, 4=FWD)

    SOURCE: fpl-video-repurposer/build_fpl_kb.py::main (line 104)
    """
    return fetch_json(BOOTSTRAP_URL)


def get_players(bootstrap: dict[str, Any] | None = None) -> list[dict[str, Any]]:
    """Return a lightweight player list from bootstrap.

    Each entry contains:
        id, first_name, second_name, web_name, team_id,
        element_type (position), status, now_cost, selected_by_percent,
        form, expected_goals, expected_assists, expected_goal_involvements

    SOURCE: fpl-video-repurposer/build_fpl_kb.py::build_master_squad (lines 38-50)
    """
    if bootstrap is None:
        bootstrap = get_bootstrap()
    return [
        {
            "id":           e["id"],
            "first_name":   e["first_name"],
            "second_name":  e["second_name"],
            "web_name":     e["web_name"],
            "team_id":      e["team"],
            "team_code":    e.get("team_code"),
            "element_type": e["element_type"],
            "status":       e["status"],
            "now_cost":     e.get("now_cost"),
            "selected_by_percent": e.get("selected_by_percent"),
            "form":         e.get("form"),
            "expected_goals":           e.get("expected_goals"),
            "expected_assists":         e.get("expected_assists"),
            "expected_goal_involvements": e.get("expected_goal_involvements"),
        }
        for e in bootstrap["elements"]
    ]


def get_teams(bootstrap: dict[str, Any] | None = None) -> list[dict[str, Any]]:
    """Return team list from bootstrap.

    SOURCE: fpl-video-repurposer/build_fpl_kb.py::build_master_squad (lines 51-56)
    """
    if bootstrap is None:
        bootstrap = get_bootstrap()
    return [
        {
            "id":         t["id"],
            "name":       t["name"],
            "short_name": t["short_name"],
            "code":       t.get("code"),
            "strength":   t.get("strength"),
        }
        for t in bootstrap["teams"]
    ]


def get_current_gameweek(bootstrap: dict[str, Any] | None = None) -> int | None:
    """Return the current (next unfinished) gameweek number, or None if season over.

    SOURCE: New helper based on bootstrap events array.
    """
    if bootstrap is None:
        bootstrap = get_bootstrap()
    for event in bootstrap.get("events", []):
        if event.get("is_current"):
            return event["id"]
    for event in bootstrap.get("events", []):
        if event.get("is_next"):
            return event["id"]
    return None


# ---------------------------------------------------------------------------
# Fixtures
# ---------------------------------------------------------------------------

def get_fixtures(gameweek: int) -> list[dict[str, Any]]:
    """Return fixture list for a specific gameweek.

    SOURCE: fpl-video-repurposer/build_fpl_kb.py::build_next_fixture_map
    """
    return fetch_json(FIXTURES_URL.format(gameweek=gameweek))


def get_fixture_difficulty_map(
    gameweek: int,
    teams: list[dict[str, Any]] | None = None,
    bootstrap: dict[str, Any] | None = None,
) -> dict[int, int]:
    """Return {team_id: fixture_difficulty} for a given gameweek.

    Difficulty is opponent strength (2-5 scale). Returns 3 (neutral) for
    double gameweeks or blank gameweeks.

    SOURCE: captaincy-showdown/src/services/captaincyDataService.ts::getFixtureDifficulty
    """
    if teams is None:
        teams = get_teams(bootstrap)

    team_strength: dict[int, int] = {t["id"]: t.get("strength", 3) for t in teams}
    fixtures = get_fixtures(gameweek)

    difficulty_map: dict[int, int] = {}
    for fix in fixtures:
        home_id = fix["team_h"]
        away_id = fix["team_a"]
        difficulty_map[home_id] = team_strength.get(away_id, 3)
        difficulty_map[away_id] = team_strength.get(home_id, 3)

    return difficulty_map


# ---------------------------------------------------------------------------
# Player element summary (match-by-match history)
# ---------------------------------------------------------------------------

def get_player_history(player_id: int) -> dict[str, Any]:
    """Return a player's element summary (history, fixtures, history_past).

    SOURCE: New helper; not yet used by any project but needed for chat interface.
    """
    return fetch_json(ELEMENT_SUMMARY_URL.format(player_id=player_id))

File: packages/fpl-api-client/typescript/src/fplClient.ts
/**
 * fpl-api-client · packages/fpl-api-client/typescript/src/fplClient.ts
 * ======================================================================
 * Browser/Node HTTP client for the official FPL API.
 *
 * SOURCE:  Extracted and generalised from:
 *   - captaincy-showdown/src/services/http.ts         (empty file — was a placeholder)
 *   - captaincy-showdown/src/services/captaincyDataService.ts::getFixtureDifficulty
 *   - captaincy-showdown/src/utils/csvPathConfig.ts   (path conventions)
 *
 * REPLACES (do NOT delete originals until migration is approved):
 *   - captaincy-showdown/src/services/http.ts         → remove & import from here
 *
 * CONSUMERS AFTER MIGRATION:
 *   - captaincy-showdown/src/services/captaincyDataService.ts
 *   - fpl-platform/apps/fpl-chat (LLM chat backend)
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface FplPlayer {
  id: number;
  first_name: string;
  second_name: string;
  web_name: string;
  team: number;        // team_id
  team_code: number;
  element_type: number; // 1=GKP 2=DEF 3=MID 4=FWD
  status: string;
  now_cost: number;
  selected_by_percent: string;
  form: string;
  expected_goals: string;
  expected_assists: string;
  expected_goal_involvements: string;
}

export interface FplTeam {
  id: number;
  name: string;
  short_name: string;
  code: number;
  strength: number;
}

export interface FplBootstrap {
  elements: FplPlayer[];
  teams: FplTeam[];
  events: FplEvent[];
}

export interface FplEvent {
  id: number;
  name: string;
  deadline_time: string;
  is_current: boolean;
  is_next: boolean;
  finished: boolean;
}

export interface FplFixture {
  id: number;
  event: number;
  team_h: number;
  team_a: number;
  team_h_difficulty: number;
  team_a_difficulty: number;
  started: boolean;
  finished: boolean;
}

// ---------------------------------------------------------------------------
// Endpoint constants
// ---------------------------------------------------------------------------

const BOOTSTRAP_URL = "https://fantasy.premierleague.com/api/bootstrap-static/";
const FIXTURES_URL = (gw: number) =>
  `https://fantasy.premierleague.com/api/fixtures/?event=${gw}`;

// ---------------------------------------------------------------------------
// Fetch helper (browser-native fetch)
// ---------------------------------------------------------------------------

async function fetchJson<T>(url: string): Promise<T> {
  const resp = await fetch(url);
  if (!resp.ok) {
    throw new Error(`FPL API error ${resp.status} for ${url}`);
  }
  return resp.json() as Promise<T>;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Return the full bootstrap-static payload (players + teams + events). */
export async function getBootstrap(): Promise<FplBootstrap> {
  return fetchJson<FplBootstrap>(BOOTSTRAP_URL);
}

/** Return the current gameweek number (is_current=true), or null. */
export function getCurrentGameweek(bootstrap: FplBootstrap): number | null {
  const curr = bootstrap.events.find(e => e.is_current);
  if (curr) return curr.id;
  const next = bootstrap.events.find(e => e.is_next);
  return next?.id ?? null;
}

/** Return fixtures for a specific gameweek. */
export async function getFixtures(gameweek: number): Promise<FplFixture[]> {
  return fetchJson<FplFixture[]>(FIXTURES_URL(gameweek));
}

/**
 * Return a Map<teamId, fixtureDifficulty> for a gameweek.
 *
 * SOURCE: captaincy-showdown/src/services/captaincyDataService.ts::getFixtureDifficulty
 */
export async function getFixtureDifficultyMap(
  gameweek: number,
  teams: FplTeam[]
): Promise<Map<number, number>> {
  const teamStrength = new Map(teams.map(t => [t.id, t.strength ?? 3]));
  const fixtures = await getFixtures(gameweek);
  const map = new Map<number, number>();
  for (const fix of fixtures) {
    map.set(fix.team_h, teamStrength.get(fix.team_a) ?? 3);
    map.set(fix.team_a, teamStrength.get(fix.team_h) ?? 3);
  }
  return map;
}


